### PR TITLE
gateway-api: Add support for TLSRoute

### LINKED ISSUE #21927: CFP: Support TLSRoute in Cilium Gateway API
## Cilium Feature Proposal
        
The first cut for Gateway API was done in https://github.com/cilium/cilium/pull/21749, however, we have a couple of follow-up tasks to support more features from Gateway API upstream.

This is to track the work to support [TLSRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TLSRoute)

## Goals

To implement [TLSRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TLSRoute) into Cilium's existing GatewayAPI implementation in a way that enables further L7 routes to be painlessly implemented.

This would allow routing of TLS traffic without termination of the encryption but allowing it to be routed based on the SNI as seen in the handshake with the GatewayAPI specification. 

## Proposed Solution

**DRAFT by Maartje**
### Envoy

The `HTTPRoute` uses an integration to convert the route to an Envoy rule that is installed via a CEC rule. This is analogue to how ingress does it using many of the same translation "middleware" code. 
`TLSRoute` is very similar to `HTTPRoute` with the difference that it only checks 1 part of the header to route: the SNI. 
A lot of the same logic could be re-used in here to make this work. 

Envoy has support to filter traffic based on SNI in the `serverNames` fillter chain match for TLS with the `TlsInspector` filter, the same logic is used to chose the route to terminate TLS to in the `HTTPRoute`. From there on we can use a TCP proxy to forward the TLS including the initial handshake to the backend service which uses the same discovery mechanism as HTTPRoute/Ingress. 

```yaml
   resources:
    - '@type': type.googleapis.com/envoy.config.listener.v3.Listener
      filterChains:
      - filterChainMatch:
          serverNames:
          - abc.example.com
          transportProtocol: tls
        filters:
        - name: envoy.filters.network.tcp_proxy
          typedConfig:
            '@type': type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
            cluster: gateway-conformance-infra/tls-backend:443
            statPrefix: gateway-conformance-infra/tls-backend:443
      listenerFilters:
      - name: envoy.filters.listener.tls_inspector
        typedConfig:
          '@type': type.googleapis.com/envoy.extensions.filters.listener.tls_inspector.v3.TlsInspector
```

This implementation can build on top of all work already done for HTTPRoute and try to consolidate many helper functions to accept both routes. However Go's type system and the TLSRoute only being v1alpha might make this harder.
Many of this work then can be used again for the GRPCRoute

## Tasks

- [x] Add required reconcile functions to validate TLSRoutes
  - [x] Validate against invalid SNIs
  - [x]  If both the Listener and TLSRoute have specified hostnames, and none match with the criteria above, then the TLSRoute is not accepted
- [x] Add support for Gateways to be TLS and Envoy to be configured acordingly
- [x] Add a translator module for TLSRoute rules to Envoy rules
- [x] Handle edge cases according to spec (TBD)
  - [x]  multi protocol gateways
  - [x] If both the Listener and TLSRoute have specified hostnames, any TLSRoute hostnames that do not match the Listener hostname MUST be ignored (includes *. handling)
- [x] Validate required CRDs will be installed
- [x] Pass conformance tests
- [ ] Awesome documentation!

### PR DESCRIPTION
**Pull request by meyskens**
This PR adds support for the TLSRoute feature of GatewayAPI.
The aim is to re-use as much of the logic we made for HTTPRoute before, as well as to improve it to add future L7 gateway types without too much of a hassle.

Under the hood it generates a CEC which sets a TLS handshake filter to capture the SNI and route based on that. From there on Envoy is configured to proxy on TCP level so the endpoint gets the full handshake and the proxy is basically in TLS passthrough mode where it cannot decrypt the traffic as is defined in the TLSRoute spec.

It also upgrades is to v0.6.1 in conformance tests to enable conformance testing to be done for TLSRoute.

Fixes: https://github.com/cilium/cilium/issues/21927

```release-note
Add TLSRoute support to GatewayAPI
```

### CODE REVIEW

**Review by kaworu — APPROVED** [review]
Helm changes LGTM.

File: operator/pkg/gateway-api/helpers.go
Comment (youngnick) [inline operator/pkg/gateway-api/helpers.go]:
```suggestion
// isAllowed returns true if the provided Route is allowed to attach to given gateway
```

File: operator/pkg/gateway-api/route_checks.go
Comment (youngnick) [inline operator/pkg/gateway-api/route_checks.go:24]: It _may_ be better to also do ReferenceGrant checking here as well as the other places it gets done; I don't think there are any conformance checks for TLSRoute ReferenceGrants yet, but this may fail.

Maybe a TODO so we can grep for ReferenceGrant later if we need to? Non-blocking for now though.

File: operator/pkg/model/translation/translator.go
Comment (youngnick) [inline operator/pkg/model/translation/translator.go:126]: I agree, letting Go manage the capacity here is fine.

**Review by youngnick — APPROVED** [review]
LGTM, but I think this needs a @sayboras review as well at least before merging. My comments are non-blocking.

File: operator/pkg/gateway-api/route_checks.go
Comment (sayboras) [inline operator/pkg/gateway-api/route_checks.go]: nit: backendValidatonFunc -> backendValidationFunc

File: operator/pkg/gateway-api/tlsroute.go
Comment (sayboras) [inline operator/pkg/gateway-api/tlsroute.go]: HTTPRoute -> TLSRoute

File: operator/pkg/gateway-api/tlsroute.go
Comment (sayboras) [inline operator/pkg/gateway-api/tlsroute.go]: same as above

File: operator/pkg/model/translation/translator.go
Comment (sayboras) [reply operator/pkg/model/translation/translator.go:126]: recently, we actually update a lot of code to pre-allocate slice size for performance reason.

File: .github/workflows/conformance-gateway-api.yaml
Comment (sayboras) [inline .github/workflows/conformance-gateway-api.yaml:170]: can you update the gatewayapi version in this file as well ? 

https://github.com/cilium/cilium/blob/2f11767a22547f8fac98d4c2b36b326524db8934/.github/workflows/conformance-gateway-api.yaml#L28

File: operator/pkg/model/translation/translator.go
Comment (sayboras) [inline operator/pkg/model/translation/translator.go]: this function can be replaced by common utility added recently

https://github.com/cilium/cilium/pull/25069

File: operator/pkg/model/translation/translator.go
Comment (sayboras) [inline operator/pkg/model/translation/translator.go]: same as above

**Review by sayboras — APPROVED** [review]
Good stuffs :astonished:, I have a couple of below comments, but it's not a blocker. 

Can you also add a small user-facing docs with TLSRoute example?

File: operator/pkg/gateway-api/route_checks.go
Comment (viktor-kurchenko) [inline operator/pkg/gateway-api/route_checks.go]: NIT: this and next line can be replaced with `return ctrl.Result{}, true, nil` and then you can get rid of `found` flag.

**Review by viktor-kurchenko — APPROVED** [review]
LGTM.

**Comment by meyskens** [comment]
> Can you also add a small user-facing docs with TLSRoute example?

Happy to but that might be better in a follow-up PR to not block this one on documentation improvements

**Comment by sayboras** [comment]
test-1.26-net-next hits https://github.com/cilium/cilium/issues/25119, which will be quarantined as part of https://github.com/cilium/cilium/pull/25236 soon.

**Review by tklauser — APPROVED** [review]
CI and vendoring changes LGTM. I haven't looked at the other bits in detail as I figure these were already covered by other reviewers.

**Comment by sayboras** [comment]
> test-1.26-net-next hits #25119, which will be quarantined as part of #25236 soon.

Reviews are in, CIs are all green except the above flake, marking this ready to merge.

**Comment by joestringer** [comment]
/test-1.26-net-next

**Comment by youngnick** [comment]
Looks like that test failure is #24966, which is a duplicate of #15455. Not related, so merging.

**Pull request merged**
